### ISSUE TITLE
How should Gemm operator in ONNX be converted?

### ISSUE THREAD

**Issue by imai-lm** [dmlc/nnvm#194]
In #186, for coverting Gemm operator of ONNX, @zhreshold wrote:
> So basically Gemm == sym.dense(use_bias=False), adding it to the _convert_map should work.

But for me it seems difficult; NNVM's dense operator is `Y=XW^T+b`, where the 2nd operator `W` will be transposed at the calculation. However, ONNX's Gemm operator is actually `Y = alpha * A * B + beta * C`,  where the 2nd operator `B` should not be transposed (unless the attribute `transB` is non-zero).

Is there a better way to solve this? Or should I add another operator like `nnvm.symbol.dense_transpose` that calculates `Y=XW+b`, for example?

**Comment by tqchen** [comment]
We can just insert a transpose op for the cast that is not covered by dense for now. Since it is not yet bottleneck, we can have internal optimizer to pattern match and optimize layer

**Comment by tqchen** [comment]
For background, most efficient implementation of gemm actually would do transposition inside  before doing the dot, so there is likely no perf loss if we do it right.

**Comment by imai-lm** [comment]
Well, the current implementation of `nnvm.frontend.onnx()` is based on [`_convert_map`](https://github.com/dmlc/nnvm/blob/master/python/nnvm/frontend/onnx.py#L102), where the conversion is defined with one-on-one basis (i.e. one operator is converted to one operator). Should we change this architecture a bit and modify it from upper levels?

**Comment by zhreshold** [comment]
@imai-lm Yes, you can modify the one-to-one basis and take/return symbols directly just like this #175

**Comment by imai-lm** [comment]
OK, I will try it anyway.

**Comment by zhreshold** [comment]
@imai-lm How's your progress? I'm going to have 0.2 supported.

**Comment by masahi** [comment]
@imai-lm @zhreshold I'm interested in this too. I need a solid onnx loading support to test our new tvm backend.

**Comment by imai-lm** [comment]
@masahi Well, I have been coding it, but it doesn't work with my model, so far. I'm looking into it so please wait for a moment.

**Comment by imai-lm** [comment]
I'm still working, but you can watch it if you are interested: https://github.com/dmlc/nnvm/compare/master...imai-lm:add-onnx-gemm-spike

**Comment by masahi** [comment]
@imai-lm thanks. Is this gemm op used by gemm based conv ops?

**Comment by imai-lm** [comment]
@masahi I'm trying to convert `Gemm` op to `nnvm.symbol.dense` with additional `transpose` ops.

**Comment by masahi** [comment]
@imai-lm yeah, I want to know how this gemm op is going to be used. 
Say, I want to load some pytorch-trained model which uses gemm based conv internally. To load this model, gemm op mush be supported, correct?

**Comment by imai-lm** [comment]
@masahi Yes, I'm checking my implementation with a pytorch-trained model, too. And actually it seems that FC layers are represented with Gemm in ONNX, generally.

**Comment by masahi** [comment]
@imai-lm Ah, that makes sense. Because FC layer IS just gemm.

**Comment by imai-lm** [comment]
Now I have encountered the error like below:
```
[14:10:35] /Users/imai/nnvm/dmlc-core/include/dmlc/./logging.h:308: [14:10:35] src/top/nn/nn.cc:44: Operator dense(units=9216, use_bias=True, name=dense0) expects weight's shape to be [9216,0], but got [4096,9216]
```
The coverted model (I printed out using `nnvm.symbol.debug_str()` is the following (sorry if it is too long):
```
Symbol Outputs:
	output[0]=dense2(0)
Variable:input_0
Variable:param_0
--------------------
Op:conv2d, Name=conv2d0
Inputs:
	arg[0]=input_0(0) version=0
	arg[1]=param_0(0) version=0
Attrs:
	channels=64
	dilation=(1, 1)
	groups=1
	kernel_size=(11, 11)
	padding=(2, 2)
	strides=(4, 4)
	use_bias=False
Variable:param_1
--------------------
Op:broadcast_add, Name=broadcast_add0
Inputs:
	arg[0]=conv2d0(0)
	arg[1]=param_1(0) version=0
--------------------
Op:relu, Name=relu0
Inputs:
	arg[0]=broadcast_add0(0)
--------------------
Op:max_pool2d, Name=max_pool2d0
Inputs:
	arg[0]=relu0(0)
Attrs:
	ceil_mode=False
	padding=(0, 0)
	pool_size=(3, 3)
	strides=(2, 2)
Variable:param_2
--------------------
Op:conv2d, Name=conv2d1
Inputs:
	arg[0]=max_pool2d0(0)
	arg[1]=param_2(0) version=0
Attrs:
	channels=192
	dilation=(1, 1)
	groups=1
	kernel_size=(5, 5)
	padding=(2, 2)
	strides=(1, 1)
	use_bias=False
Variable:param_3
--------------------
Op:broadcast_add, Name=broadcast_add1
Inputs:
	arg[0]=conv2d1(0)
	arg[1]=param_3(0) version=0
--------------------
Op:relu, Name=relu1
Inputs:
	arg[0]=broadcast_add1(0)
--------------------
Op:max_pool2d, Name=max_pool2d1
Inputs:
	arg[0]=relu1(0)
Attrs:
	ceil_mode=False
	padding=(0, 0)
	pool_size=(3, 3)
	strides=(2, 2)
Variable:param_4
--------------------
Op:conv2d, Name=conv2d2
Inputs:
	arg[0]=max_pool2d1(0)
	arg[1]=param_4(0) version=0
Attrs:
	channels=384
	dilation=(1, 1)
	groups=1
	kernel_size=(3, 3)
	padding=(1, 1)
	strides=(1, 1)
	use_bias=False
Variable:param_5
--------------------
Op:broadcast_add, Name=broadcast_add2
Inputs:
	arg[0]=conv2d2(0)
	arg[1]=param_5(0) version=0
--------------------
Op:relu, Name=relu2
Inputs:
	arg[0]=broadcast_add2(0)
Variable:param_6
--------------------
Op:conv2d, Name=conv2d3
Inputs:
	arg[0]=relu2(0)
	arg[1]=param_6(0) version=0
Attrs:
	channels=256
	dilation=(1, 1)
	groups=1
	kernel_size=(3, 3)
	padding=(1, 1)
	strides=(1, 1)
	use_bias=False
Variable:param_7
--------------------
Op:broadcast_add, Name=broadcast_add3
Inputs:
	arg[0]=conv2d3(0)
	arg[1]=param_7(0) version=0
--------------------
Op:relu, Name=relu3
Inputs:
	arg[0]=broadcast_add3(0)
Variable:param_8
--------------------
Op:conv2d, Name=conv2d4
Inputs:
	arg[0]=relu3(0)
	arg[1]=param_8(0) version=0
Attrs:
	channels=256
	dilation=(1, 1)
	groups=1
	kernel_size=(3, 3)
	padding=(1, 1)
	strides=(1, 1)
	use_bias=False
Variable:param_9
--------------------
Op:broadcast_add, Name=broadcast_add4
Inputs:
	arg[0]=conv2d4(0)
	arg[1]=param_9(0) version=0
--------------------
Op:relu, Name=relu4
Inputs:
	arg[0]=broadcast_add4(0)
--------------------
Op:max_pool2d, Name=max_pool2d2
Inputs:
	arg[0]=relu4(0)
Attrs:
	ceil_mode=False
	padding=(0, 0)
	pool_size=(3, 3)
	strides=(2, 2)
--------------------
Op:reshape, Name=reshape0
Inputs:
	arg[0]=max_pool2d2(0)
Attrs:
	shape=(10, 9216)
--------------------
Op:dropout, Name=dropout0
Inputs:
	arg[0]=reshape0(0)
Attrs:
	rate=0.5
Variable:param_10
--------------------
Op:transpose, Name=transpose0
Inputs:
	arg[0]=param_10(0) version=0
Attrs:
	axes=(1, 0)
--------------------
Op:transpose, Name=transpose1
Inputs:
	arg[0]=transpose0(0)
Attrs:
	axes=(1, 0)
Variable:param_11
--------------------
Op:dense, Name=dense0
Inputs:
	arg[0]=dropout0(0)
	arg[1]=transpose1(0)
	arg[2]=param_11(0) version=0
Attrs:
	units=9216
	use_bias=True
--------------------
Op:relu, Name=relu5
Inputs:
	arg[0]=dense0(0)
--------------------
Op:dropout, Name=dropout1
Inputs:
	arg[0]=relu5(0)
Attrs:
	rate=0.5
Variable:param_12
--------------------
Op:transpose, Name=transpose2
Inputs:
	arg[0]=param_12(0) version=0
Attrs:
	axes=(1, 0)
--------------------
Op:transpose, Name=transpose3
Inputs:
	arg[0]=transpose2(0)
Attrs:
	axes=(1, 0)
Variable:param_13
--------------------
Op:dense, Name=dense1
Inputs:
	arg[0]=dropout1(0)
	arg[1]=transpose3(0)
	arg[2]=param_13(0) version=0
Attrs:
	units=4096
	use_bias=True
--------------------
Op:relu, Name=relu6
Inputs:
	arg[0]=dense1(0)
Variable:param_14
--------------------
Op:transpose, Name=transpose4
Inputs:
	arg[0]=param_14(0) version=0
Attrs:
	axes=(1, 0)
--------------------
Op:transpose, Name=transpose5
Inputs:
	arg[0]=transpose4(0)
Attrs:
	axes=(1, 0)
Variable:param_15
--------------------
Op:dense, Name=dense2
Inputs:
	arg[0]=relu6(0)
	arg[1]=transpose5(0)
	arg[2]=param_15(0) version=0
Attrs:
	units=4096
	use_bias=True
```

**Comment by masahi** [comment]
@imai-lm I'd love to help you, but I'm full with my job and rocm backend development for tvm.
@zhreshold can you take a look?

**Comment by imai-lm** [comment]
Apparently, the inferred shape of weight doesn't match the actual shape coming from ONNX model... I'm a bit confused now.

**Comment by imai-lm** [comment]
I guess the code [here](https://github.com/dmlc/nnvm/blob/b107c3e61c30659b8726e768a1bc7811a75c0be0/python/nnvm/frontend/onnx.py#L367):
`channels = out_shapes[0][0]`
may be:
`channels = out_shapes[0][1]`
This suits my situation (though it produces another error). Is there any test code that tests this part?
